### PR TITLE
TestPluginCache and other dependencies refactoring

### LINKED ISSUE #2634: Assembly caching in Test Platform
## Description
Provide a way to efficiently get necessary types from assemblies

### PR DESCRIPTION
**Pull request by Sanan07**
## Description
Refactoring of TestPluginCache and other classes responsible for test discovery and test execution

Fix : #2634

### CODE REVIEW

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -12,6 +13,7 @@
 [assembly: AssemblyCopyright("� Microsoft Corporation. All rights reserved.")]
 [assembly: AssemblyProduct("Microsoft.TestPlatform.Extensions.BlameDataCollector")]
 [assembly: AssemblyTrademark("")]
+[assembly: InterestingTypes("Microsoft.TestPlatform.Extensions.BlameDataCollector.BlameLogger")]
```
[the remark this answers is not in the document]
Comment (Sanan07) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Properties/AssemblyInfo.cs]: Yes, I was also thinking about correct name and could not figure out better one. I will try to get something more relevant

[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Properties/AssemblyInfo.cs]: I was thinking `TestPlatformExtensionType`, `TestPlatformExtensionTypesToInspect` but it all sounds clunky.

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs [hunk 2]
```diff
@@ -196,14 +192,11 @@ private void AddKnownExtensions(ref IEnumerable<string> extensionPaths)
                 return;
             }
 
-            if ((types != null) && (types.Length > 0))
+            if (types != null && types.Count() > 0)
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs:195]: .Any()?

File: src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    [AttributeUsage(AttributeTargets.Assembly, Inherited = false, AllowMultiple = false)]
+    [CLSCompliant(false)]
+    public sealed class TypesToLoadAttribute : Attribute
+    {
+        public TypesToLoadAttribute(params Type[] types)
+        {
+            Types = types;
+        }
+
+        public IEnumerable<Type> Types { get; }
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs:21]: Why are you using ienumerable instead of `Type[]` ?

**Review by nohwnd — COMMENTED** [review]
Do we have some measurements on this? I would like to see it do it's magic. How much time should this save? And how to best test it?

**Comment by Sanan07** [comment]
@nohwnd Do you mean measurements of using custom attributes? If assembly has custom attributes, on average it takes 1.5-2ms to get that "interesting" types, compared to 30ms which takes calling `GetTypes()` method on assembly. So having that custom attributes on several assemblies (for now 5 assemblies and will add to MSTestAdapter as well) can save us around ~170 - 200ms.

File: src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs (hunk 3 above)
Comment (Sanan07) [reply src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs:21]: Do not remember why, will fix it.

**Comment by nohwnd** [comment]
Yeah, just wanted the benefit to be documented here for future reference. :)

File: src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs [hunk 4]
```diff
@@ -18,6 +17,6 @@ public TypesToLoadAttribute(params Type[] types)
             Types = types;
         }
 
-        public IEnumerable<Type> Types { get; }
+        public Type[] Types { get; }
```
Comment (Haplois) [inline src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs:20]: This is no longer immutable.

**Review by Haplois — APPROVED** [review]
Looks good, but I have a nit-picking comment.

File: src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs (hunk 4 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.ObjectModel/Utilities/TypesToLoadAttribute.cs:20]: IEnumerable, is also not immutable isn't it? But does it matter? How would this collection change to cause problems? We would need a static ctor to find it and modify it on assembly load?

**Pull request merged**